evm_loader/program/src/executor/state.rs

```diff
@@ -284,18 +285,23 @@ impl<'a, B: AccountStorage> Database for ExecutorState<'a, B> {
         Ok(self.backend.code_size(from_address).await)
     }
 
-    async fn code_hash(&self, from_address: Address, chain_id: u64) -> Result<[u8; 32]> {
-        use solana_program::keccak::hash;
+    async fn code_hash(&self, address: Address, chain_id: u64) -> Result<[u8; 32]> {
+        // https://eips.ethereum.org/EIPS/eip-1052
+        // https://eips.ethereum.org/EIPS/eip-161
 
-        for action in &self.actions {
-            if let Action::EvmSetCode { address, code, .. } = action {
-                if &from_address == address {
-                    return Ok(hash(code).to_bytes());
-                }
-            }
+        let code = self.code(address).await?;
+
+        if !code.is_empty() {
+            return Ok(keccak::hash(&code).to_bytes());
         }
 
-        Ok(self.backend.code_hash(from_address, chain_id).await)
+        if self.nonce(address, chain_id).await? == 0 && self.balance(address, chain_id).await? == 0
+        {
+            return Ok(<[u8; 32]>::default());
+        }
+
+        // equal to "c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470"
+        return Ok(keccak::hash(&[]).to_bytes());
```

Comment:
As @anton-lisanin mentioned [here](https://github.com/neonlabsorg/neon-evm/pull/253#discussion_r1448863426), I suggest first checking for `nonce == 0 && balance == 0`, then computing the hash, so the code becomes even simpler. Something like:
```Rust
        let code_hash = {
            let address = self.stack.pop_address()?;

            // https://eips.ethereum.org/EIPS/eip-1052
            // https://eips.ethereum.org/EIPS/eip-161

            // An account with code has non-zero nonce
            if backend.nonce(address, self.chain_id).await? == 0
                && backend.balance(address, self.chain_id).await? == 0
            {
                <[u8; 32]>::default()
            } else {
                let code = backend.code(address).await?;
                keccak::hash(&code).to_bytes()
            }
        };
```

Reply:
While the suggestion implementation seems correct, does that mean that we would use up more instructions in the common case where the account is a program? Is that the common case? Are we okay with that?

Reply:
I don't know the common code path in this method.

Reply:
@anton-lisanin we require your expertise :)

Reply:
@anton-lisanin mentioned in the daily that indeed we expect this to be called mostly for accounts that **do** have code associated with them.